src/duk_debugger.c

```diff
@@ -1448,6 +1448,49 @@ DUK_LOCAL void duk__debug_handle_detach(duk_hthread *thr, duk_heap *heap) {
 	DUK__SET_CONN_BROKEN(thr, 0);  /* not an error */
 }
 
+DUK_LOCAL void duk__debug_handle_apprequest(duk_hthread *thr, duk_heap *heap) {
+	DUK_D(DUK_DPRINT("debug command AppRequest"));
+
+	if (heap->dbg_request_cb != NULL) {
+		duk_context *ctx = (duk_context *) thr;
+		duk_idx_t nrets;
+		duk_idx_t nvalues = 0;
+		duk_idx_t old_top;
+		duk_idx_t top, idx;
+
+		/* Read tvals from the message and push them onto the valstack,
+		 * then call the request callback to process the request.
+		 */
+		old_top = duk_get_top(ctx);  /* save stack top */
+		while (duk_debug_peek_byte(thr) != DUK_DBG_MARKER_EOM) {
+			duk_debug_read_tval(thr);  /* push to stack */
+			nvalues++;
+		}
+
+		/* Request callback should push values for reply to client onto valstack */
+		nrets = heap->dbg_request_cb(heap->dbg_udata, ctx, nvalues);
+		if (nrets > 0) {
+			DUK_ASSERT(duk_get_top(ctx) >= old_top + nrets);
+
+			/* Reply with tvals pushed by request callback */
+			duk_debug_write_byte(thr, DUK_DBG_MARKER_REPLY);
+			top = duk_get_top(ctx);
+			for (idx = top - nrets; idx < top; idx++) {
+				duk_debug_write_tval(thr, DUK_GET_TVAL_POSIDX(ctx, idx));
```

Comment:
Ok I see, you'd suggest leaving the assert there before the if-check.

I guess that makes sense because the callback is violating its requirements.

It's not always an indication of value stack being compromised so it *may* be safe to continue. If we knew it's *not safe* to continue I'd prefer just using the assert.